packages/Search/src/details/DTK_DetailsPriorityQueue.hpp

```diff
@@ -36,55 +36,110 @@ template <typename T, typename Compare = Less<T>>
 class PriorityQueue
 {
   public:
-    using SizeType = size_t;
+    using IndexType = ptrdiff_t;
+    using ValueType = T;
+    using ValueCompare = Compare;
 
     KOKKOS_FUNCTION PriorityQueue() = default;
 
-    KOKKOS_INLINE_FUNCTION bool empty() const { return _size == 0; }
+    KOKKOS_FUNCTION bool empty() const { return _size == 0; }
+
+    KOKKOS_FUNCTION IndexType size() const { return _size; }
+
+    KOKKOS_FUNCTION void clear() { _size = 0; }
+
+    KOKKOS_FUNCTION ValueCompare value_comp() const { return _compare; }
 
     template <typename... Args>
-    KOKKOS_FUNCTION void push( Args &&... args )
+    KOKKOS_INLINE_FUNCTION void push( Args &&... args )
     {
-        // ensure the queue is not already full
+        // ensure the heap is not already full
         assert( _size < _max_size );
 
-        // construct the element to compare to those in the queue
-        T elem( std::forward<Args>( args )... );
-
-        // find position of the new element in the sorted array
-        // COMMENT: could consider implementing a binary search here
-        SizeType pos;
-        for ( pos = 0; pos < _size; ++pos )
-            if ( !_compare( _queue[pos], elem ) )
+        // add the element to the bottom level of the heap
+        IndexType pos = _size;
+        T elem{std::forward<Args>( args )...};
+
+        // perform up-heap operation
+        while ( pos > 0 )
+        {
+            // compare the added element with its parent
+            // if they are in correct order, stop
+            // if not, swap them and continue
+            IndexType const parent = ( pos - 1 ) / 2;
+            if ( !_compare( _heap[parent], elem ) )
                 break;
+            _heap[pos] = _heap[parent];
+            pos = parent;
+        }
 
-        // move memory to make room for it
-        for ( SizeType tmp = _size; tmp > pos; --tmp )
-            _queue[tmp] = _queue[tmp - 1];
-
-        // insert the new element
-        _queue[pos] = elem;
+        _heap[pos] = elem;
 
-        // update the size of the queue
+        // update the size of the heap
         ++_size;
     }
 
     KOKKOS_INLINE_FUNCTION void pop()
     {
+        // ensure that the heap is not empty
         assert( _size > 0 );
+
+        // replace the root with the last element on the last level
+        IndexType pos = 0;
+
+        // perform down-heap operation
+        while ( true )
+        {
+            // compare the new root with its children
+            // if they are in the correct order, stop
+            // if not, swap the element with one of its children and continue
+            IndexType const left_child = 2 * pos + 1;
+            IndexType const right_child = 2 * pos + 2;
+            IndexType next_pos = _size - 1;
+            for ( IndexType child : {left_child, right_child} )
+                if ( child < _size - 1 &&
+                     _compare( _heap[next_pos], _heap[child] ) )
+                    next_pos = child;
```

Comment:
I feel like there is some potential for optimization here. If we check right child first, and `_compare` is "greater than" operation, do we still need to check `left_child`? I.e., would something like this work?
```c++
if (right_child < _size-1 && _compare(_heap[next_pos], _heap[right_child])
  next_pos = right_child;
else if (left_child < _size-1 && _compare(_heap[next_pos], _heap[left_child])
  next_pos = left_child;
```